riscv/src/register/macros.rs

```diff
@@ -600,3 +600,979 @@ macro_rules! clear_pmp {
+                range: [$field_start : $field_end],
+                default: $default_variant,
+                $($variant = $value,)+
+            },
+        );
+
+        $crate::read_only_csr_field!(
+            $ty,
+            $(#[$field_doc])*
+            $field: [$field_start : $field_end],
+            $field_ty,
+        );
+    };
+
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident: [$field_start:literal : $field_end:literal],
+     $field_ty:ident$(,)?
+    ) => {
+         impl $ty {
+             paste::paste! {
+                 $(#[$field_doc])+
+                 #[inline]
+                 pub fn $field(&self) -> $field_ty {
+                     self.[<try_ $field>]().unwrap()
+                 }
+
+                 $(#[$field_doc])+
+                 #[inline]
+                 pub fn [<try_ $field>](&self) -> $crate::result::Result<$field_ty> {
+                    let max_width = core::mem::size_of_val(&self.bits) * 8;
+
+                     if $field_end < max_width && $field_start < $field_end {
+                         let value = $crate::bits::bf_extract(
+                             self.bits,
+                             $field_start,
+                             $field_end - $field_start + 1,
+                         );
+
+                         $field_ty::from_usize(value).ok_or($crate::result::Error::InvalidFieldVariant {
+                             field: stringify!($field),
+                             value,
+                         })
+                     } else {
+                        Err($crate::result::Error::IndexOutOfBounds {
+                            index: $field_start,
+                            min: $field_start,
+                            max: $field_end,
+                        })
+                     }
+                 }
+             }
+         }
+    };
+}
+
+/// Defines field accesor functions for a write-only CSR type.
+#[macro_export]
+macro_rules! write_only_csr_field {
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident,
+     $(#[$try_field_doc:meta])+
+     $try_field:ident,
+     bit: $bit:literal$(,)?) => {
+        impl $ty {
+            $(#[$field_doc])+
+            #[inline]
+            pub fn $field(&mut self, $field: bool) {
+                self.$try_field($field).unwrap();
+            }
+
+            $(#[$try_field_doc])+
+            #[inline]
+            pub fn $try_field(&mut self, $field: bool) -> $crate::result::Result<()> {
+                let max_width = core::mem::size_of_val(&self.bits) * 8;
+
+                if $bit < max_width {
+                    self.bits = $crate::bits::bf_insert(self.bits, $bit, 1, $field as usize);
+                    Ok(())
+                } else {
+                    Err($crate::result::Error::IndexOutOfBounds {
+                        index: $bit,
+                        min: 0,
+                        max: max_width,
+                    })
+                }
```

Comment:
We can do this as a compile-time check, right? we do know the value of `max_width` and `$bit`, so we can do a compile-time check to verify that everything is OK and just use the `pub fn $field` function, as we are sure that it will not fail.

Reply:
When working on this, I was considering the safety of the `bits` functions. There should be `try_` variants of the `bf_insert` and `bf_extract` functions, because both of these functions can implicitly panic.

We would use the `checked_shl` function instead of `<<` operator.

What would the compile-time check look like here, as an example?

Reply:
The `$bit` param is fixed, right? For example, the bit number 3 of a register. Thus, we can add something like [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=cd70233e4269326d495c04dddda11f63) to check at compile time that `$bit` is within the expected bounds. In this way, we don't need `try_` methods